### PR TITLE
[persist/expr] Refine the output of `EXPLAIN WITH(filter_pushdown)`

### PR DESCRIPTION
**Pull request by bkirwi**
Previously, we'd consider a function-call expression "pushdownable" if any of its arguments were, and the function itself was monotonic etc. This ends up being too generous: for example, `col + hash(col)` is not predictable even if the left-hand side is, because the right-hand side could be some arbitrary value.

Getting this right requires distinguishing between expressions that don't depend on the column ranges because they're constants, and expressions that don't depend on the column ranges because they're complicated functions... so we have to change up the summary type here as well.

### Motivation

Previously unreported bug.

In brief, we discovered a case where the EXPLAIN output was misleadingly suggesting that a particular filter would be pushed down where in practice it was never useful; see the [internal conversation](https://materializeinc.slack.com/archives/C053EPHMU05/p1698676211422729) for more details.

### Tips for reviewer

I don't feel that excited about the naming of the new types / variants; happy to take suggestions.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by danhhz — COMMENTED** [review]
Seems sane, but I think I need to leave the review to compute!

File: src/expr/src/interpret.rs [hunk 1]
```diff
@@ -898,41 +898,74 @@ impl<'a> Interpreter for ColumnSpecs<'a> {
...
-/// second argument is pushdownable, `/` is not monotone on the right-hand side).
+/// In practice, we internally need to distinguish between expressions that are trivially predicable
+/// (because they're constant) and expressions that depend on the column ranges themselves.
+/// See the [TraceSummary] variants for those distinctions, and [TraceSummary::pushdownable] for
+/// the overall assessment.
 #[derive(Debug)]
 pub struct Trace;
 
+/// A summary type for the [Trace] interpreter.
+// NB: the ordering here is important!
```
Comment (aalexandrov) [inline src/expr/src/interpret.rs:910]: nit: can you make this a rustdoc comment and explain which part of the code relies on the ordering at the moment?

File: src/expr/src/interpret.rs [hunk 2]
```diff
@@ -898,41 +898,74 @@ impl<'a> Interpreter for ColumnSpecs<'a> {
...
+#[derive(Copy, Clone, Debug, PartialOrd, PartialEq, Ord, Eq)]
+pub enum TraceSummary {
+    /// The expression is constant: we can evaluate it without any runtime information.
+    Constant,
+    /// The expression depends on runtime information, but in "predictable" way... ie. if we know
+    /// the range of possible values for all columns and unmaterializable functions, we can
+    /// predict the possible values of the output.
+    Dynamic,
+    /// The expression depends on runtime information in an unpredictable way.
```
Comment (aalexandrov) [inline src/expr/src/interpret.rs:919]: nit: expand what you mean by "an unpredictable way". IIUC this entails that knowledge about value ranges of the inputs cannot be used to infer the value range of the output.

File: src/expr/src/interpret.rs [hunk 3]
```diff
@@ -898,41 +898,74 @@ impl<'a> Interpreter for ColumnSpecs<'a> {
...
     }
 
     fn literal(&self, _result: &Result<Row, EvalError>, _col_type: &ColumnType) -> Self::Summary {
-        false
+        TraceSummary::Constant
     }
 
     fn unmaterializable(&self, _func: &UnmaterializableFunc) -> Self::Summary {
-        false
+        TraceSummary::Dynamic
```
Comment (aalexandrov) [inline src/expr/src/interpret.rs:960]: Why is this changing from `false` to `Dynamic` and not to `Unknown`? In which cases can you interpret an unmaterializable function to a specific value range?

File: test/sqllogictest/filter-pushdown.slt [hunk 4]
```diff
@@ -311,3 +311,63 @@ Source materialize.public.t
   pushdown=((timestamp_to_mz_timestamp(case when (#0 = 0) then (#1 - 1 day) else (#1 - 2 days) end) < mz_now()))
 
 EOF
+
+# Regression test: should not report pushdown when one branch might throw an exception.
```
Comment (aalexandrov) [inline test/sqllogictest/filter-pushdown.slt:315]: One branch of what? I think this is referring to the `CASE` branches below.

File: src/expr/src/interpret.rs [hunk 5]
```diff
@@ -941,30 +974,34 @@ impl Interpreter for Trace {
...
-            .any(|pushdownable_arg| pushdownable_arg && pushdownable_fn)
+            .map(|pushdownable_arg| pushdownable_arg.apply_fn(pushdownable_fn))
+            .max()
+            .unwrap_or(TraceSummary::Constant)
     }
 
     fn cond(&self, cond: Self::Summary, then: Self::Summary, els: Self::Summary) -> Self::Summary {
-        cond || (then || els)
+        // We don't actually need to be able to predict the condition precisely to predict the output,
+        // since it's just a boolean.
```
Comment (aalexandrov) [inline src/expr/src/interpret.rs:1002]: Is it because it's a boolean or because you can union the ranges of the two branches for a conservative estimate?

**Review by aalexandrov — APPROVED** [review]
LGTM.

Some minor nits based on which maybe you can improve comments before merging this.

File: src/expr/src/interpret.rs (hunk 3 above)
Comment (bkirwi) [reply src/expr/src/interpret.rs:960]: We do this for `mz_now()`... in the persist source we can bound the range based on the as-of of the query. (See [here](https://github.com/MaterializeInc/materialize/blob/f8c5f3a2939f7e3bdc3aa0909826b5fd50744ae8/src/expr/src/interpret.rs#L682-L690) for how that looks on the mz-expr side.) We could in theory fill in constant values for some of the other functions, but in practice they get compiled away by the time the expression reaches us...

File: src/expr/src/interpret.rs (hunk 5 above)
Comment (bkirwi) [reply src/expr/src/interpret.rs]: The latter's probably the more obvious explanation... switched it over.

**Comment by bkirwi** [comment]
Thanks for the suggestions! Made the changes and queued up the merge.

**Pull request merged**